[mac] document and adjust to standards based timing concepts

fgrandel opened the pull request:
Hi OpenThread community and above all @edmont , I'm heading over to you guys to adjust OpenThread to a recent unification of timing concepts in the Zephyr IEEE 802.15.4 implementation.

In a nutshell this small change set proposes to align some of the timing sensitive aspects of OpenThread even more closely with concepts from the underlying standards. Unification and closer specification of timing along original IEEE 802.15.4 concepts enables implementation of additional non-Thread IEEE 802.15.4 features in Zephyr (namely TSCH).

Most of this change set is documentation - only a small change is being made to the interpretation of the TX timestamp that seems not to be particularly restricted in OpenThread so far. I propose that we make use of this leeway to jointly move closer to standards-conforming timing primitives.

AFAIK, Zephyr and its derivatives and radio drivers (notably the nRF Connect SDK and nRF radio driver) are the only platforms that depend on this code and those have already made the corresponding changes on their side.

See https://github.com/zephyrproject-rtos/zephyr/issues/59245 for more context as well as @edmont 's regression test result with real hardware based on this change set.

edmont reviewed (approved):
Thanks @fgrandel, I definitely agree on having well defined timing concepts and it makes total sense to have the same reference for reception and transmission timestamps.

Notice that apart from Zephyr there are some other platforms affected by this change, at least https://github.com/openthread/ot-efr32 and https://github.com/openthread/ot-nxp.

Also @EskoDijk could you take a look in the context of https://github.com/openthread/ot-ns?

fgrandel commented:
> Notice that apart from Zephyr there are some other platforms affected by this change

@edmont How should we go about these platforms?

edmont commented:
> > Notice that apart from Zephyr there are some other platforms affected by this change
> 
> @edmont How should we go about these platforms?

Lets allow @jwhui to ping relevant maintainers.

Irving-cl commented on src/core/thread/csl_tx_scheduler.hpp:
```diff
@@ -105,10 +105,53 @@ class CslTxScheduler : public InstanceLocator, private NonCopyable
-        uint16_t  mCslPhase;            ///< The time when the next CSL sample will start.
-        TimeMilli mCslLastHeard;        ///< Time when last frame containing CSL IE was heard.
-        uint64_t  mLastRxTimestamp;     ///< Time when last frame containing CSL IE was received, in microseconds.
+        uint16_t  mCslPeriod;           ///< CSL sampled listening period between channel samples in units of 10 symbols (160 microseconds).
```
Maybe change like this?
```suggestion
        uint16_t  mCslPeriod;           ///< CSL sampled listening period between the beginning point of 2 consecutive channel samples in units of 10 symbols (160 microseconds).
```

fgrandel replied:
@Irving-cl Thanks for your review! Inserted "consecutive" but left out the "beginning point" part as a period can by definition be measured between any corresponding points, not just the beginning of the samples, do you agree?

Irving-cl commented on src/core/thread/csl_tx_scheduler.hpp:
```diff
@@ -105,10 +105,53 @@ class CslTxScheduler : public InstanceLocator, private NonCopyable
+         */
+        uint16_t  mCslPhase;
+        TimeMilli mCslLastHeard;        ///< Radio clock time when last frame containing CSL IE was heard.
+        uint64_t  mLastRxTimestamp;     ///< Radio clock time when last frame containing CSL IE was received, see `Mac::RxFrame::GetTimestamp`.
```
This field is defined in `CslTxScheduler::ChildInfo` and only used for CSL, not the one obtained by`Mac::RxFrame::GetTimestamp`. We can probably do enhancement by merging this with `mCslLastHeard`. I'll check if we can do this. Let's keep the original comment here.

fgrandel replied:
Fixed.

Irving-cl reviewed (approved):
LGTM 👍

romacdon commented:
@hemanth-silabs / @suveshpratapa - could you two review this PR, thanks.

fgrandel commented:
I believe I fixed the prettify problem - not sure how to fix the version problem - could you point me to some docs that give me a hint?

jwhui commented:
> I believe I fixed the prettify problem - not sure how to fix the version problem - could you point me to some docs that give me a hint?

You simply need to increment the value relative to the existing value on main.

fgrandel commented:
> You simply need to increment the value relative to the existing value on main.

Ok, read through the script and found out where it looks for the version. Then bumped OPENTHREAD_API_VERSION in include/openthread/instance.h

fgrandel commented:
@jwhui - All checks pass now except for Simulation 1.1. But is that error really due to my change? It seems that the pending timestamp is missing from the TLV set in the child update request. I'd assume that this wouldn't be influenced by my change?

Irving-cl commented:
@fgrandel  I've re-triggered the checks. They pass now.

abtink commented on src/core/thread/csl_tx_scheduler.hpp:
```diff
@@ -101,14 +101,59 @@ class CslTxScheduler : public InstanceLocator, private NonCopyable
+         * nextTxTimestamp
+         *         = nextTmh - phrDuration
+         *         = lastRxTimestamp + 160us * (n * cslPeriod + cslPhase)
+         */
```
Thanks for adding detailed explanation. 👍

abtink reviewed (approved):
LGTM. Thanks.

EskoDijk commented on include/openthread/platform/time.h:
```diff
@@ -52,17 +52,34 @@ extern "C" {
+ * Get the current time in microseconds referenced to a continuous monotonic
+ * local radio clock (64 bits width).
+ *
+ * The radio clock SHALL not wrap during the device's uptime. Implementations
```
In case of a normative requirement the sentence should be:

```suggestion
 * The radio clock SHALL NOT wrap during the device's uptime. Implementations
```

fgrandel replied:
Fixed.

EskoDijk commented on include/openthread/platform/time.h:
```diff
@@ -52,17 +52,34 @@ extern "C" {
+ *
+ * The radio clock SHALL not wrap during the device's uptime. Implementations
+ * SHALL therefore identify and compensate for internal counter overflows. The
+ * clock does not have a defined epoch nor SHALL it introduce any continuous or
```
Correction on normative requirement language:

```suggestion
 * clock does not have a defined epoch and SHALL NOT introduce any continuous or
```

fgrandel replied:
Fixed.

EskoDijk commented on include/openthread/platform/time.h:
```diff
@@ -52,17 +52,34 @@ extern "C" {
+ * Get the current estimated worst case accuracy (maximum ± deviation from the
+ * nominal frequency) of the local radio clock in units of PPM.
+ *
+ * Implementations MAY estimate this value based on current operating conditions
```
There's some information missing here: in case the implementation does not estimate the value but returns a fixed value, this value MUST be the worst-case accuracy over all possible foreseen operating conditions (temperature, pressure, etc) of the implementation.

fgrandel replied:
Fixed.

EskoDijk commented on src/core/mac/sub_mac.hpp:
```diff
@@ -667,7 +667,7 @@ class SubMac : public InstanceLocator, private NonCopyable
                               // reception.
     uint16_t    mCslPeerShort;      // The CSL peer short address.
-    TimeMicro   mCslSampleTime;     // The CSL sample time of the current period.
+    TimeMicro   mCslSampleTime;     // The CSL sample time of the current period referred to the local radio clock.
```
```suggestion
    TimeMicro   mCslSampleTime;     // The CSL sample time of the current period referenced to the local radio clock.
```

EskoDijk replied:
Or "relative to" is even more clear (but in that case the same wording should be used in other places as well)

fgrandel replied:
You are right "relative to" is a common way to express what I mean here (see PTP spec for example). Fixed everywhere.

fgrandel commented:
@Irving-cl @abtink @jwhui Do you think this is ready for merge now? The other requested reviewers seem to be missing in action. Please be aware that this PR blocks dependent issues in Zephyr.

EskoDijk commented:
@fgrandel Not missing in action - there wasn't any deadline requested I think ;)

fgrandel commented:
@EskoDijk Yes, I'm not sure how typical expected response times are in the OT community - I assumed 2 weeks would be fine. Can you approve the PR from your side? I think I addressed all your concerns.

EskoDijk commented on src/core/thread/csl_tx_scheduler.hpp:
```diff
@@ -101,14 +101,59 @@ class CslTxScheduler : public InstanceLocator, private NonCopyable
+         *       expectations.
+         *     * The CSL Transmitter SHALL start transmitting the first symbol
+         *       of the preamble of the frame to transmit at the preamble time
+         *       T_pa = (CSL-Phase-Time – 192 μs) (that is, CCA must be
```
@abtink : any policy for special characters in source? I recall that usually "us" or "microseconds" is used which avoids the mu symbol.

fgrandel replied:
Fixed.

abtink replied:
I don't believe we have any specific guideline or policy for this, but as @EskoDijk  mentioned, I think we have not used them in the past. It might be helpful to add a guideline on this later. I personally do like them, as they make the text look nicer, but I am not sure if they may cause any issues on different editors and/or older toolchains.

fgrandel replied:
@EskoDijk @abtink I did a search on OT source code and the mu symbol appears in a few places, but the us/microsecond notation is much more frequent (and I changed my docs accordingly). I think the symbol was introduced because I shamelessly copied it from the standard. But no matter what your decision, I'll adapt. No strong feelings re one or the other on my side.

abtink replied:
Thanks for checking it. Good idea to go with the more commonly used model (use usec/microseconds).

EskoDijk reviewed (changes requested):
see comments. Maybe the change on the txWindow calculations can be done in a separate PR? As it modifies the calculation and breaks any platforms that use a non-time-scheduled Tx platform.

fgrandel commented on include/openthread/platform/time.h:
```diff
@@ -52,17 +52,39 @@ extern "C" {
 
 /**
- * Get the device's XTAL accuracy.
+ * Get the current estimated worst case accuracy (maximum ± deviation from the
```
Yes, you're right. Fixed.

fgrandel commented on src/core/thread/csl_tx_scheduler.cpp:
```diff
@@ -158,9 +158,10 @@ uint32_t CslTxScheduler::GetNextCslTransmissionDelay(const Child &aChild,
-    uint64_t nextTxWindow = radioNow - (radioNow % periodInUs) + (firstTxWindow % periodInUs);
+
+    /* see CslTxScheduler::ChildInfo::mCslPhase */
+    uint64_t firstTxWindow = aChild.GetLastRxTimestamp() + aChild.GetCslPhase() * kUsPerTenSymbols;
```
@EskoDijk Thank you for looking into the change at such a level of detail! This ensures that we don't overlook anything and that everyone understands what this implies. Let me try to answer your questions as precisely as I can.

> Is it intentional that kRadioHeaderShrDuration is removed here -- so that the point of reference for nextTxWindow becomes "end of last symbol of SFD" ?

Moving to a common standard based definition for TX and RX timestamps was the central motivation for this PR as those definitions were inconsistent in OT AFAICS (see detailed rationale of my choice of timestamp definition below).

> Because this changes the calculation, it also means that code below in sub_mac.cpp becomes invalid:

I don't think so. The change was introduced asymmetrically as the TX side was already based on the SFD definition (at least in Zepyhr's platform adaptation layer). I may have done it wrong or my memory may fail me - but that was the intention. @edmont You're probably in a much better place to explain the details on the OT side or see the errors in my implementation or argument.

> Maybe the change on the txWindow calculations can be done in a separate PR? As it modifies the calculation and breaks any platforms that use a non-time-scheduled Tx platform.

All the documentation is just preparative to introduce the concepts needed to define timestamps with the precision required by the IEEE 802.15.4 standard itself. The documentation describes and explains the implementation change and can therefore not be separated from it IMHO. I just did two commits to ensure that the implementation change is not "lost in the noise". Of course I may be wrong or the documentation may not reflect the implementation - so I hope such errors will be unveiled in this review.

This PR proposes a common definition for TX and RX timestamps for the following reasons:
* Such an inconsistency is difficult to reason about and easy to overlook. My intention is to reduce cognitive load and improve maintainability (better long term TCO at the price of short term refactoring effort).
* There is also a technical argument that might not (yet) be relevant to OpenThread but this can easily change in the future (as the Thread standard may not ignore highly relevant recent PHYs like UWB or SUN FSK forever). The IEEE 802.15.4 standard defines a "start of frame delimiter" that it guarantees to be consistently defined across all current and future PHYs such that MAC protocols can be based on that definition regardless of PHY details. SFD is defined as the earliest point in time at which the existence of a frame on the medium can be predicted with sufficiently high probablility and precision (including ranging requirements) for a given RF demodulation/correlation/filter technology. The SHR duration depends very much on the PHY and can therefore not be consistently used on MAC level. AFAIK this is the reason why most frame timings in the standard are relative to the "start of the first symbol after the SFD" or "end of SFD". If even more precision is required (e.g. in ranging) then the timestamp reference plane is defined to be "the local antenna" and called "RMARKER" (but it's the same basic concept). The fact that CSL also introduces timings relative to the start of the MAC header is probably to ease the implementation of the corresponding IE parsers and as you can see from my formula is not a problem as long as the PHY specific offsets cancel. Such a definition is not adequate for general timestamps independent of the IEEE 802.15.4 sub-protocol, though, and should therefore not be used outside CSL standard for general RX/TX timestamps.

I agree that it is unfortunate to fix such an inconsistency after it has been implemented by all platforms. But OTOH the longer it lives the harder it will be to fix. So IMO it's still cheaper to fix it now than in two years.

IMO the change is less than what you assume, though. It's just that the length of the SHR (which is just a simple well-defined us-precision constant for the given PHY) may have to be added in platform implementations that choose to use different timestamp definitions internally. In Zephyr we simply migrated our whole stack down to the driver level to a standard definition - but that's our choice of course. I think it's obvious that this is the right thing to do for Zephyr (which already supports all kinds of PHYs and IEEE 802.15.4 standard protocols and expects to do some work in ranging soon).

That said, I'm just offering my help here. Zephyr's decision to use standard timestamps can easily be isolated from the OT API by transforming timestamps internally to Zephyr's OT platform adaptation layer. So of course the OT community may legitimately decide to use different timestamp definition for whatever theoretical or practical reasons (which would be a valid result of the review!).

fgrandel replied:
@EskoDijk Fixed the SHR offset for platforms that don't support timed TX as you suggested.

EskoDijk replied:
@fgrandel Ok thanks for pointing out the motivation. Most other (active) participants I think agreed with the idea to unify timestamp references. But I'm not sure that a one-type-of-time-reference approach is possible. Below some thoughts about that.  Maybe we could agree that any variable called "timestamp" uses the end-of-last-symbol-of-SFD time reference, while any variable using "window" (win) in its name uses the "when the radio is turned on or off" time reference?  If agreed in this way then I can check details of the code to see if it's all correct. And I can verify against my simulation platform to check that the CSL timings are correctly calculated.

Thoughts:

The header `include/openthread/platform/radio.h` introduces also the concept of "window" in `otPlatRadioReceiveAt` and this window concept comes back in other parts of code like `GetNextCslTransmissionDelay` and `SubMac::HandleCslTimer(void)`  (abbreviated 'win' there).  So the concept is used for the receiver (time that radio is on, see e.g. fig 6-63 in 802.15.4-2015) and also for the transmitter - although for that case it refers more to a single timestamp, that targets a time that lies within the receiver's window.

This relates directly to the 802.15.4 concept of the "channel sample". The start of this channel sample (period) is called channel sample time, and that is a timestamp that refers to the time that the radio is turned on (i.e. referencing the start of first symbol of SHR, from the receiver's perspective.) 
So in my mind the "channel sample" as defined by 802.15.4 is equivalent to the "window" that OpenThread introduces for the receiver. 

In my mind, to remain well aligned to the standards, we need to keep the same concept and not change it to "Rx window start is end of last symbol of SFD".  That would be more confusing I think. If that reasoning is correct we have to live with multiple methods of time reference, because the standard itself is also doing this.

One question here: I'm not sure if the current PR aims to change the "receive window start time" (`aStart`) parameter reference time, in `otPlatRadioReceiveAt`, from its current time interpretation of "start of window" to "end of last symbol of SFD" ?  If yes that would imply some more code changes are needed in the calls to `otPlatRadioReceiveAt`.

At the moment, on the CSL receiver, in `SubMac::HandleCslTimer(void)` the variable `winStart` is not a timestamp that is pointing to the end of last symbol of SFD. Rather it points to the "start of the window - when the CSL receiver switches on the radio".
The reason for this is simple: the code for platforms-without-receive-time-support needs to switch on the radio in Rx at this time so it's not so handy or intuitive I think to use another timestamp reference. The timer needs to trigger at the time that the radio must be switched on; and not later than that. And "window start" refers to the 802.15.4 concept of "channel sample" i.e. start of the channel sample.

fgrandel replied:
> In my mind, to remain well aligned to the standards, we need to keep the same concept and not change it to "Rx window start is end of last symbol of SFD". 

@EskoDijk Agreed! (This and everything else you write in that comment)

I never intended to say anything about RX windows or channel samples in this PR other than documenting existing realities or pointing out the simple change in offset required when deriving the existing RX windows from a slightly changed TX timestamp. I also agree that SFD is relevant for timestamping only.

The only thing that I intend to change in this PR is the TX timestamp because it deviates from the standard AND from the RX timestamp definition. (Timestamps as opposed to RX windows exactly as you define them.) If I did say anything that results in a change to the resulting RX window definition, it is unintended and should be pointed out and fixed.

EskoDijk replied:
Ok thanks, good to hear - so we can keep the "window" term for the receiver's channel sample then I think; and also move the timestamps to the agreed reference point. And the CSL timer can trigger in advance, as needed for the specific platform. With this in mind I can do another review and then test with the simulation platform! (hopefully by tomorrow)

EskoDijk commented:
Sorry one question I forgot to ask earlier: why are the references made to the radio's local clock in the file `time.h` , `otPlatTimeGet()`, while there is also a time function `otPlatRadioGetNow()` more directly within `radio.h` ?

Shouldn't the function template in `radio.h` have similar restrictions as the one in `time.h` ? I wonder why exactly is the one in `time.h` also mentioning "radio" there because at first sight this seems to be only for a time service of the platform, independent of the radio functions. 

In my simulation platform, the `otPlatRadioGetNow()` by the way internally just calls `otPlatTimeGet()`. So it could be that these are just used interchangably but I'm not sure how it was intended.

EskoDijk commented on include/openthread/platform/radio.h:
```diff
@@ -764,7 +778,8 @@ otError otPlatRadioReceive(otInstance *aInstance, uint8_t aChannel);
  *
  * @param[in]  aChannel   The radio channel on which to receive.
- * @param[in]  aStart     The receive window start time, in microseconds.
+ * @param[in]  aStart     The receive window start time relative to the local
```
@fgrandel  As discussed about the "window" definition, I wanted to make this more clear. Proposing the below clarification for the parameters:

```
/**
 * Schedule a radio reception window at a specific time and duration.
 *
 * @param[in]  aChannel   The radio channel on which to receive.
 * @param[in]  aStart     The receive window start time relative to the local radio clock, see `otPlatTimeGet`.
 *                        The window start time is when the radio is on and ready to receive the first symbol of 
 *                        the SHR.
 * @param[in]  aDuration  The receive window duration, in microseconds. The window end time is when the radio is 
 *                        turned off (in case it is not actively receiving a frame at that moment).
 *
 * @retval OT_ERROR_NONE    Successfully scheduled receive window.
 * @retval OT_ERROR_FAILED  The receive window could not be scheduled.
 */
otError otPlatRadioReceiveAt(otInstance *aInstance, uint8_t aChannel, uint32_t aStart, uint32_t aDuration);
```

Is this also what you had in mind? By using this time definition we avoid having to change various CSL code. An alternative here is to use the `mTimestamp` reference instead for `aStart`, but that's more impactful.

EskoDijk replied:
Sorry, I may be fully wrong here. Although the original text nicely refers to the "window" concept (related to CSL sample) , if I look at the CSL code it seems to be assuming another form of timestamp reference. (e.g. it mentions "MHR" still.)

Was it the intention to use `otPlatRadioReceiveAt` with the exact same time reference (SFD based) as `mTimestamp` ? In that case the header could look like:

```
/**
 * Schedule a radio reception window at a specific time and duration.
 *
 * @param[in]  aChannel   The radio channel on which to receive.
 * @param[in]  aStart     The receive window start time relative to the local radio clock, see `otPlatTimeGet`.
 *                        The (SFD based) frame time reference is the same as for `mTimestamp`.
 * @param[in]  aDuration  The receive window duration, in microseconds. The time `aStart + aDuration` refers to the 
 *                        latest possible frame that can still be reliably detected within the window.
 *
 * @retval OT_ERROR_NONE    Successfully scheduled receive window.
 * @retval OT_ERROR_FAILED  The receive window could not be scheduled.
 */
otError otPlatRadioReceiveAt(otInstance *aInstance, uint8_t aChannel, uint32_t aStart, uint32_t aDuration);
```

EskoDijk replied:
Or better, `aDuration` could simply refer to the time that the radio is on.

fgrandel replied:
> Is this also what you had in mind?

Yes, that's my understanding. @edmont Do you agree? I added you text just slightly re-formulated using SHALL/SHOULD terminology. See whether you like it.

fgrandel replied:
> Was it the intention to use `otPlatRadioReceiveAt` with the exact same time reference (SFD based) as `mTimestamp` ?

No, IMO neither MAC header nor SFD make for good RX window definitions on PHY/L1 level. Your initial comment and definition (the one that I also used now) represents exactly my understanding of a low-level RX window and I believe this is how the method is actually being implemented right now - at least in Zephyr.

Why not MAC header? "receive at" is below MAC level (L1/PHY). CSL channel samples are part of a MAC protocol (one of many IEEE 802.15.4 MAC sub-protocols). A generic "receive at" feature should not know anything about MAC level. It is true that CSL phase refers to start-of-MAC but that is very CSL specific and happens nowhere else in the standard. This definition should be converted to low level "receive at" in the MAC.

Why not SFD? Defining an RX window relative to SFD only works for very specific PHY configurations. IMO that would make the actual start of the RX window completely non-deterministic, as even within one PHY (e.g. SUN FSK) there may be several SHR lengths. And if your receiver supports several PHYs at the same time then you're completely lost.

IMO SFD is for RX/TX timestamps (including timed TX) - not more not less.

edmont replied:
> @edmont Do you agree?

I do. Thank you @fgrandel and @EskoDijk for taking the effort of making all these APIs parameters much more clear.

fgrandel commented:
> why are the references made to the radio's local clock in the file `time.h` , `otPlatTimeGet()`, while there is also a time function `otPlatRadioGetNow()` more directly within `radio.h` ?

@EskoDijk Good catch! Short answer: I just happened to stumble upon otPlatTimeGet() first because that's the only clock used in Zephyr's platform abstraction layer. otPlatRadioGetNow() is just an alias. But that is of course an oversight on my side. I'll fix that.

That being said, I believe that the OT clock architecture is not very stringent. If you check where the different clocks are used, I doubt the implementations work as independently as you suggest. The platform clock is used in several places where probably either the radio clock should be used instead or a synchronized version of the two would be needed. We're currently fixing this class of bugs in Zephyr which is where this PR originates from. But that is a big issue and definitely out of scope for this simple PR.

fgrandel commented:
> I'll fix that.

Fixed.

EskoDijk commented:
@fgrandel  I've created a branch here: https://github.com/EskoDijk/openthread/tree/pr-csl 
with a few more edits and clarifications. You can use that in this PR if needed.

From what I can see, with the current code, it's not really PHY-independent in any case. (e.g. submac / csl tx scheduler). Though with the changes it's better prepared for future PHYs.

fgrandel commented:
> it's not really PHY-independent in any case.

True. But now it is a little more of an implementation issue and a little less of an API/design issue. This PR doesn't try to solve all problems. ;-)

fgrandel commented:
> @fgrandel I've created a branch here: https://github.com/EskoDijk/openthread/tree/pr-csl with a few more edits and clarifications. You can use that in this PR if needed.

Thank you so much for investing such a lot of effort. Very much appreciated. I need to run now but I'll have time later today to do a diff and integrate your suggestions.

EskoDijk commented:
> @EskoDijk Thank you so much for investing such a lot of effort. Very much appreciated. I need to run now but I'll have time later today to do a diff and integrate your suggestions. As you can see your comments have made this a much better quality PR.

Just added a commit with some more updates; with a fix for the timestamp time reference in CSL debug log printing. There may be more coming; this was based on CSL testing with my platform which is not finished yet.

EskoDijk commented:
> 
> Just added a commit with some more updates; with a fix for the timestamp time reference in CSL debug log printing. There may be more coming; this was based on CSL testing with my platform which is not finished yet.

I probably need to recheck everything - still seeing some definition issues. The `aCslSampleTime` in `otPlatRadioUpdateCslSampleTime()` probably needs to reference just the start of the sample window, i.e. start of SHR. It's not a *timestamp* but just a *time*.

Also `radio.hpp` has very similar comments/parameters to `radio.h` , but aren't updated by this PR, maybe we should just leave those be for now. These don't interface to the platform directly.

fgrandel commented:
> I probably need to recheck everything - still seeing some definition issues.

@EskoDijk No problem - we've decoupled this PR from our internal issue by now so please take your time. I propose you recheck everything and once you're confident that you have completed, let me know and I'll review the changes in your own branch and integrate them. Thorough review is good.

EskoDijk commented:
@fgrandel Ok, I'm done now with review and testing on the simulation platform.
For convenience I've also created a PR here https://github.com/fgrandel/openthread/pull/1  to track the differences between your code and my updates on it.

Main change is to document that `otPlatRadioUpdateCslSampleTime()` uses the (now default) timestamp reference of end of SFD. This is based on how the existing code used this timestamp, and on how `mCslSampleTime` is used in sub_mac. Probably here using the default time reference helps to keep the code understandable. 

Note that the SHR time duration was already part of the "Ahead" time component of the CSL sample window; this is now better documented.

fgrandel commented:
> Main change is to document that `otPlatRadioUpdateCslSampleTime()` uses the (now default) timestamp reference of end of SFD. This is based on how the existing code used this timestamp, and on how `mCslSampleTime` is used in sub_mac. Probably here using the default time reference helps to keep the code understandable.

To be honest, I don't like that change. The spec clearly defines CSL timing in terms of the start of MAC header, see my explanation in https://github.com/openthread/openthread/pull/9322#discussion_r1291207899 about the difference between low level PHY-independent timestamps (i.e. SFD/RMARKER timestamping) as opposed to the special case of CSL (plus the lengthy inline documentation of Thread 1.3.1 wrt CSL timing and the derivation of corresponding timestamps introduced in this PR).

I would therefore not introduce any change to the definition of MAC-level CSL timing in this PR, it's out of scope. So if at all it should be changed then this should be kept in a separate PR. It's not what I intended in this PR. This PR is exclusively about introducing SFD-based timestamps _where the standard does not explicitly deviate_ from them.

fgrandel commented:
@EskoDijk Oh, and of course I have nothing against closing this PR in favor of another PR that you base on top of mine with additional changes and then merge that one. I just don't have time to go that deep into CSL timing specific to OT. It's much beyond "scratching my own itch".

EskoDijk commented on src/core/mac/sub_mac.cpp:
```diff
@@ -471,7 +471,7 @@ void SubMac::StartCsmaBackoff(void)
                     kCcaSampleInterval - kCslTransmitTimeAhead)
```
I would expect the following:

```suggestion
                    kCcaSampleInterval - kCslTransmitTimeAhead - kRadioHeaderShrDuration)
```

Reason: only if the compound TxDelay time is in the future, we need to use the timer.  In other cases the "Transmit without delay" case should be chosen.

fgrandel replied:
Fixed. :-)

EskoDijk commented:
@fgrandel 

> To be honest, I don't like that change. The spec clearly defines CSL timing in terms of the start of MAC header

Right, overlooked that part. This is indeed the reference point in the spec for CSL. Let's use that! I updated my platform code to it and it seems to work ok.

EskoDijk reviewed (changes requested):
All looks ok now (any further improvements per discussion can be done in a future PR!), except for one comment where a change seems needed.

EskoDijk reviewed (approved):
Thanks; all looks good now.

fgrandel commented:
@Irving-cl @abtink @jwhui Ready for merge?

romacdon commented:
> @hemanth-silabs / @suveshpratapa - could you two review this PR, thanks.

@suveshpratapa / @hemanth-silabs - just reminding you of this PR that needs to be reviewed.  Thanks.

hemanth-silabs reviewed (approved):
LGTM 👍

jwhui reviewed (approved):
Thanks! 👍🏼
